### ISSUE TITLE
CI/CD infrastructure for hubs that is automated, tested, and parallelized 

### ISSUE THREAD

**Issue by choldgraf** [2i2c-org/infrastructure#879]
_This is a pitch for a multi-sprint project - it uses a slightly different structure so that we can test out how it feels!_

### Problem statement

We have some basic hub deployment infrastructure set up. However this is missing a few important pieces of functionality:

- We cannot deploy hubs in parallel, which means that one hub can bottleneck all others, and that updating one hub means updating all other hubs
- Our testing infrastructure can be flaky, which makes it hard to distinguish a "true failure" from a "false failure"
- Our configuration lumps all hubs into a single config file, which makes it harder to find and refer to one hub's configuration.

### Proposed solution

We should improve our hub CI/CD infrastructure with a few specific focus areas:

- Hubs should be deployed in parallel with one another in CI (not the deployer)
- We should have automated testing infrastructure that will prevent hub deployment if it fails
- Configuration should be done via a GitHub repository, and deployment should be automatically triggered

Here's a rough idea of how this could work:

- Each cluster has a folder that contains configuration for each hub on the cluster
- Each hub gets its own configuration file, which is a valid and standalone Z2JH config.
  - Maybe there is another shared configuration file that each hub can draw from for common settings across hubs.
- Adding a new hub's configuration will trigger the deployment of a JupyterHub with that configuration on that cluster.
- We deploy staging hubs for each hub. We run some kind of test on the staging hub for common workflows, and if it fails then the production hub for that community does not deploy.
- We have a "source of truth" for the hubs that we're running that is easy to glance through, which provides important information needed to maintain them, as well as a rough status of each.

### Guidance for implementation

A few things that *must* exist in the final product:

- **Canary deployments**: Deploying a hub should be stopped if an automated test does not pass.

**Things to test for**

A few ideas to include in testing infrastructure:

- Unauthorized users cannot log in
- Authorized users can spin up an interactive server properly
- Run a notebook designed for the base environment top to bottom without errors
- Run a notebook designed for a Dask Hub environment top to bottom without errors

**No gos / out-of-scope**

- This should not become a deep dive into deploying Kubernetes on different cloud providers, unless we *really* think there's something crucial to solve there that would enable the deployment of hubs. Instead we should just assume there's a k8s cluster available and focus this on deploying hubs.

**Related issues**

These are related issues, and some (or all) may be in-scope to close as part of this project

- [ ] https://github.com/2i2c-org/infrastructure/issues/586

#### 2022-01-19

We had a team meeting today to agree on a few next steps for @sgibson91 and @consideRatio to tackle some of this. Here's a rough idea of what we'd like to get done:

- [x] @sgibson91 and @consideRatio (+ others if interested) schedule a time to speak and come up with a proposal for the configuration structure that we want.
  - Ask "what features are we hoping / planning to support with 2i2c?" "What configuration structure will allow us to support those features?"
- [x] Share the proposal with others and decide next steps
- [x] https://github.com/2i2c-org/infrastructure/issues/857
- [x] https://github.com/2i2c-org/infrastructure/issues/937
- [x] https://github.com/2i2c-org/infrastructure/issues/279
- [x] https://github.com/2i2c-org/infrastructure/issues/818

I believe that @sgibson91 is championing this with @consideRatio providing support and collaboration!

**Comment by sgibson91** [comment]
> We should have automated testing infrastructure that will prevent hub deployment if it fails

I think this needs some clarification. Currently, we run automated tests _after_ we have deployed an upgrade to a hub. So this statement could either mean:

- Chained deployments were staging is deployed first followed by production-level hubs. Failure to deploy to staging halts deployment to the productions hubs. OR
- Some other form of testing we run before the deployment step, e.g., validating the helm template

> **Canary deployments:** Deploying a hub should be stopped if an automated test does not pass.

Again, clarification here. This should probably say "deploying to production-level hubs is halted if deployment to staging hubs fail"

> Adding a new hub's configuration will trigger the deployment of a JupyterHub with that configuration on that cluster.

This already happens since the CI/CD runs the deployer and the deployer uses the `--install` flag to create a deployment if it doesn't already exist. But we often don't do it in practice because, if for some reason deployment fails, we can currently only see logs locally, not in CI/CD, as secrets can be leaked.

---

I think it would also be cool to add `test-this-pr` to this project to enable testing on staging hubs from a PR. This fleshes out the **canary deployments**.

---

I am very keen to champion this project :smile:

**Comment by consideRatio** [comment]
I'd like to be an assistant on this project!

This work requires changes to the _deployer configuration file schema_. I suggest two phases of this work.

1. __Phase 2: Refactoring of the deployer script's configuration file schema__
  1. Suggest what changes are required (@sgibson91 @choldgraf)
  2. Review, iterate, and agree on changes (@2i2c-org/tech-team)
  3. Implement refactoring changes (@sgibson91)
2. __Phase 2: Update CI system__
  - Add helm template validation (#279) (@consideRatio does it, @sgibson91 assists)
  - Optimize deployment ordering (#818) (@sgibson91 does it, @consideRatio assists)
3. __At any time__
  - I think #586 can be worked on at any time independently from the other tasks.

**Comment by choldgraf** [comment]
Related to #908

We had an incident where unauthorized users could access a hub because the hub's configuration did not have `allowed_users` set (thus, JupyterHub allowed *all* users).

We should add some kind of validation for certain configuration that we know needs to be set. Maybe this is part of #279 ?

**Comment by choldgraf** [comment]
### Update

In our latest team meeting it seemed like @sgibson91 and @consideRatio were interested in focusing their efforts on this work for their next major focus area. To begin with they were going to work on https://github.com/2i2c-org/infrastructure/issues/857, and this effort will likely create new ideas for what else needs to be done in the CI/CD.

@consideRatio and @sgibson91 - how do you feel about scoping this effort for two weeks, and then doing a re-assessment and adjusting our scope then?

**Comment by sgibson91** [comment]
I think let's see how far we get in a month and then decide if what's left is worth another cycle?

**Comment by choldgraf** [comment]
That sounds good to me - but let's define what work we want to accomplish in that month. Of all the improvements that we could make towards CI/CD, what is a reasonable subset to shoot for? We can adjust over time as we learn more, but we'll be more effective if we have a target to aim towards.

Can @sgibson91 and @consideRatio scope out what they think would be the most impactful and realistic given that time frame?

**Comment by sgibson91** [comment]
I think definitely:

- isolating each hub config (and the changes to the deployer that requires); and
- parallelising CI

**Comment by choldgraf** [comment]
### Next steps

- [ ] @sgibson91 and @consideRatio (+ others if interested) schedule a time to speak and come up with a proposal for the configuration structure that we want.
  - Ask "what features are we hoping / planning to support with 2i2c?" "What configuration structure will allow us to support those features?"
- [ ] Share the proposal with others and decide next steps

**Comment by sgibson91** [comment]
https://github.com/2i2c-org/infrastructure/issues/903 is a small fix that I think would be easily tackled as part of this project too

**Comment by sgibson91** [comment]
@2i2c-org/tech-team we have outlined a proposal for the refactor of the file structure here: https://hackmd.io/@29EjZnuAQx-ME7Olgi2JXw/Sk5S7xP6F

**Comment by damianavila** [comment]
I like the proposal!! Left a tiny comment but, from the top of my head, I do not see a blocker so I would be +1 on it.

**Comment by choldgraf** [comment]
I went through and left a few comments myself as well - in general I think it looks like a pretty clean and well-structured setup. A nice improvement!

**Comment by sgibson91** [comment]
Thanks for the comments so far! If there are no strong pushbacks in the next day or so, I'll begin implementing the batches of work outlined in the document.

**Comment by sgibson91** [comment]
I am going to start working on the following todos from the hackmd in order:

- [ ] Rename `hub-templates` folder to `helm-charts`
- [ ] Migrate the `hubs.*.template` key to be `hubs.*.helm_chart` in `*cluster.yaml`
- [ ] Relocate the support chart into the `helm-charts` folder
- [ ] Rename `/config/hubs` to `/config/clusters`
- [ ] Migrate the `support.config` key to `support.helm_chart_values_files` and separate out the helm chart values from the cluster configuration
- [ ] Refactor configuration structure
  -  Migrate the `hubs.*.config` key to be `hubs.*.helm_chart_values_files` and separate the helm chart values into individual files.
  - Let each cluster have it’s own folder under `/config/clusters` that contains all the helm chart values files for each deployment
  - Migrate each `*.cluster.yaml` file to be a `cluster.yaml` file under each cluster folder in `/config/cluster`
- [ ] Add a new function to the deployer that can generate:
  - A list of helm chart values files from a given cluster name and hub name
  - the reverse
- [ ] Update deployer validate function to run `helm template`

**Comment by sgibson91** [comment]
If anyone has opinions on https://github.com/2i2c-org/infrastructure/issues/965 while I'm poking around in the deployer, I'd love to hear

**Comment by sgibson91** [comment]
**Update from a quick Slack chat with Erik:** values files should be defined relative to the location of the `cluster.yaml` file, not the location of where the deployer is invoked from

**Comment by sgibson91** [comment]
Drafting in here: https://github.com/2i2c-org/infrastructure/pull/985

**Comment by sgibson91** [comment]
How to deal with these sections is going to cause a headache. We don't know ahead of time which files contain the placeholder, so I can't see a way around this other than reading in every file in the list until you find the placeholder.

https://github.com/2i2c-org/infrastructure/blob/701e4a78c46e55c90df453a56b2c25461ac20110/deployer/utils.py#L49-L106

**Comment by consideRatio** [comment]
I'm not sure about "staff placeholders" meaning, there is no docstring for this function :cry:. I think though, its templating configuration for various values to add/remove 2i2c staff to various lists? Maybe that should be done in logic as part of the jupyterhub startup routine as a hub.extraConfig entry (like a jupyterhub_config.py snippet) instead?

**Comment by sgibson91** [comment]
It is replacing this:

https://github.com/2i2c-org/infrastructure/blob/701e4a78c46e55c90df453a56b2c25461ac20110/config/clusters/pangeo-hubs.cluster.yaml#L78

with the list maintained here:

https://github.com/2i2c-org/infrastructure/blob/master/config/clusters/staff.yaml

**Comment by GeorgianaElena** [comment]
>@2i2c-org/teams/tech-team it would be great to have your input on a decision like this!

Dealing with the staff situation gave me lots of headaches too :( As someone who already tried various ways of implementing this and failed a few times from various reasons (https://github.com/2i2c-org/infrastructure/pull/311), I will dare to suggest to just go with the solution that makes life easier! The placeholder implementation was such type of solution too. So, whatever makes more sense with the new refactored structure.

The script suggestion sounds good :+1: (though I remember trying and failing to implement something similar because I didn't know how to determine which authentication method was chosen (google/github) from within the `hub.extraConfig` and also the `username_pattern` seemed hard to determine :disappointed:)

**Comment by consideRatio** [comment]
Any additional information for the script in `hub.extraConfig` can be passed via the `custom` section, preferably for example `custom.2i2c.<something>` as that would help clarify that this is 2i2c that has given this custom config meaning for anyone down the line to better comprehend what's related to what.

Here is a reduced example of what I'm thinking.

```yaml
custom:
  2i2c:
    addStaffUsersToAdminUsers: true
    staffUsers:
      - GeorgianaElena
      - sgibson91
      - consideratio
    # in practice, we need to have a list of different types of accounts and
    # some logic to know what to use, perhaps by inspecting the kind of
    # c.JupyterHub.authenticator_class has been configured or via
    # via explicit declaration in `custom.2i2c.staffUserListToUse` etc
hub:
  extraConfig:
    2i2c_addStaffUsersToAdminUsers: |
      # ...
      staff_users = get_config("custom.2i2c.staffUsers", [])
      add_staff_users_to_admin_users = get_config("custom.2i2c.addStaffUsersToAdminUsers", False)
      if add_staff_users_to_admin_users:
          c.JupyterHub.admin_users.extend(staff_users)
```

**Comment by consideRatio** [comment]
> Are you suggesting we add the staff list in basehub too?

Yes I suggest both the script (`hub.extraConfig...`) and the staff list (`custom.2i2c...`) is centralized still, for example via basehub default values or via a dedicated helm chart values file that the various charts reference as one `--values <file>` to pass as part of `helm upgrade` etc.

**Comment by GeorgianaElena** [comment]
> Any additional information for the script in hub.extraConfig can be passed via the custom section, preferably for example custom.2i2c.<something>

That's a good idea! However, I still don't see how we can determine the auth type from within the script, without duplicating/moving this info in the custom section.  https://github.com/2i2c-org/infrastructure/blob/701e4a78c46e55c90df453a56b2c25461ac20110/config/clusters/2i2c.cluster.yaml#L24 

It might be that I'm missing something though

**Comment by consideRatio** [comment]
You are probably right about it being hard to decide what kind of user identity should be used. The authenticator may be against auth0 no matter what, but still different usernames etc should be used.

Since that is a configuration that I understand to be on how the deployer script automatically setups things with auth0, I suggest not entangling it with the choice of user lists anyhow which may not relate to auth0.

Here is a config structure proposal and perhaps functional example to add to the basehub default values, but to be combined with each helm chart configuring their own dedicated values that include setting `custom.2i2c.add_staff_user_ids_to_admin_users` and `custom.2i2c.add_staff_user_ids_of_type`.

While they could rely on a default value in the basehub helm chart, I think it's better that they don't to increase comprehensibility.

```yaml
custom:
  2i2c:
    # Should staff user id's be injected to the admin_users
    # configuration of the JupyterHub's authenticator by our
    # custom jupyterhub_config.py snippet as declared in hub.extraConfig?
    add_staff_user_ids_to_admin_users: false
    add_staff_user_ids_of_type: ""
    staff_github_ids:
      - choldgraf
      - consideRatio
      - damianavila
      - GeorgianaElena
      - sgibson91
      - yuvipanda
    staff_google_ids:
      - choldgraf@2i2c.org
      - erik@2i2c.org
      - damianavila@2i2c.org
      - georgianaelena@2i2c.org
      - sgibson@2i2c.org
      - yuvipanda@2i2c.org
hub:
  extraConfig:
    2i2c_add_staff_user_ids_to_admin_users: |
      add_staff_user_ids_to_admin_users = get_config("custom.2i2c.add_staff_user_ids_to_admin_users", False)
      if add_staff_user_ids_to_admin_users:
          user_id_type = get_config(f"custom.2i2c.add_staff_user_ids_of_type")
          staff_user_ids = get_config(f"custom.2i2c.staff_{user_id_type}_ids", [])
          c.JupyterHub.admin_users.extend(staff_user_ids)
```

**Comment by choldgraf** [comment]
> I will dare to suggest to just go with the solution that makes life easier! The placeholder implementation was such type of solution too. So, whatever makes more sense with the new refactored structure.

I think that this is the right approach here. I can't speak to the specific proposal because I don't understand the deployer setup well enough, but I recall that in previous conversations, we were trying to balance "what works best for 2i2c" as well as "what makes it easy for others to replicate our setup from a R2R perspective".

At this point, I think it's more important that we make our lives easier and simpler. If it means there is some 2i2c special-casing in our deploying infrastructure, I think that's fine. I think it will be better to enable "Right to Replicate" by making sure the Z2JH / TLJH docs are good, and providing others some guidance in following them, rather than by choosing our own deployment setup so that it is copy/pasteable.

Not sure if this is a helpful comment or not, but thought I'd share in case this was a decision-factor

**Comment by consideRatio** [comment]
> Or are there things that @consideRatio would like to work on next week while Sarah is gone?

I have a full plate of work things to focus on already for at least two weeks. I'm not sure on the amount of work remaining to get those PRs merged, but I'd be happy to expedite review effort.

**Comment by sgibson91** [comment]
I think I should be able to have #988 ready for review/merge today and then can push as far as I can on #985. Theoretically, if there are no more blockers, it should be doable by the end of the week.

**Comment by sgibson91** [comment]
Ok, next blocker 😅 

[This function](https://github.com/2i2c-org/infrastructure/blob/fcd83405e42847b7b322d7d91b94e84a21c0d2b1/deployer/hub.py#L479) expects to be able to read the helm chart values in order to see if the `docs_service` is enabled and generate a token for it.

https://github.com/2i2c-org/infrastructure/blob/fcd83405e42847b7b322d7d91b94e84a21c0d2b1/deployer/hub.py#L504-L511

The `generated_config` never gets explicitly written to the helm chart values file. It is written to a temporary file and passed as another values file instead.

https://github.com/2i2c-org/infrastructure/blob/fcd83405e42847b7b322d7d91b94e84a21c0d2b1/deployer/hub.py#L564-L595

**Comment by consideRatio** [comment]
I think what I'd like most, would be if we broke the DRY principle. That would help demystify the deployer script which is too magic too me overall - doing things I don't keep track of it doing.

We have planned for something like...

```yaml
hubs:
  - name: staging
    domain: staging.us-central1-b.gcp.pangeo.io
    helm_chart: daskhub
    auth0:
      enabled: false
    helm_chart_values_files:
      - staging.values.yaml
```

What I'm thinking would help a lot with comprehensibility of what is done by the deployer script, and this situation, would be if we had like...

```yaml
hubs:
  - name: staging
    domain: staging.us-central1-b.gcp.pangeo.io
    helm_chart: daskhub
    auth0:
      enabled: false
    helm_chart_values_generated:
      docs_service_api_token: true
    helm_chart_values_files:
      - staging.values.yaml
```

**Comment by sgibson91** [comment]
Is token generation not a thing we could move into a basehub tpl file? Like: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/417e84e81101f5757c5bb735bd7daf249696c45f/jupyterhub/templates/hub/_helpers-passwords.tpl

**Comment by consideRatio** [comment]
@sgibson91 ooooh yes! Actually, by not specifying an API token, one is automatically generated by Z2JH! That means, the basehub helm chart can simply reference it directly - and voila, no need to do anything by let it be accessed! :D

I'll see if I have written an example about this already in z2jh docs.

**Comment by consideRatio** [comment]
@sgibson91 hmmm... I'm confused though about who is using the token. To have a api_token for a JupyterHub registered service means that that service can be granted permission to speak with JupyterHub, but... I don't see any use of that token by the docs service - the docs service doesn't seem to use any such token...

Does it even need a token? I don't think so.

**Comment by consideRatio** [comment]
It could be that this is a legacy of a bug that was fixed, about needing an api_token specified for a service, even though it really shouldn't be needed. So, an option perhaps is to simply not specify such token at all any more if its only created for that purpose.

**Comment by GeorgianaElena** [comment]
Found the relevant PR @consideRatio https://github.com/2i2c-org/infrastructure/pull/445 I don't think we figure out why we ended up needing it :confused:

**Comment by consideRatio** [comment]
Ah, the resolution was https://github.com/jupyterhub/jupyterhub/pull/3531, available in jupyterhub version 1.4.2, which is available in z2jh 1.1.0+, and we use version: 1.1.3. We can safely just whipe the api token generation part.

Okay nice! So we just stop generating the api token part. Further, we have it written out to say... `"url": f'http://docs-service.{self.spec["name"]}',` where we specify the service by also being explicit about what namespace - but that may not be needed as long as the proxy pod is in the same namespace as the docs service. So, it can be hardcoded to `http://docs-service` i think.

And, then, suddenly we have no config related to a secret, and no config related to the namespace etc. Voila - we can hardcode a default value in basehub just like is done for configurator:

https://github.com/2i2c-org/infrastructure/blob/4baa7a7a2bcc5caf48558269ac2dbf3090301d89/helm-charts/basehub/values.yaml#L272-L279

We should not specify a `command` though for this docs_service, we just want the proxy pod to be configured to proxy traffic to the docs service.

**Comment by sgibson91** [comment]
This is amazing work - thank you @consideRatio and @GeorgianaElena!!!

Do you think the tpl files would be a good thing to implement for the [hub-health-service token](https://github.com/2i2c-org/infrastructure/blob/fcd83405e42847b7b322d7d91b94e84a21c0d2b1/deployer/hub.py#L492-L495) and the [dask-gateway api token](https://github.com/2i2c-org/infrastructure/blob/fcd83405e42847b7b322d7d91b94e84a21c0d2b1/deployer/hub.py#L520-L528)? If so, we can get rid of the `apply_hub_helm_chart_fixes` function completely which I think will greatly reduce complexity!!

**Comment by sgibson91** [comment]
I opened https://github.com/2i2c-org/infrastructure/pull/997 to hardcode the docs service

**Comment by consideRatio** [comment]
@sgibson91 by inspection and looking at ...

```python
        # Generate a token for the hub health service
        hub_health_token = hmac.new(
            secret_key, b"health-" + self.spec["name"].encode(), hashlib.sha256
        ).hexdigest()
        # Describe the hub health service
        generated_config.setdefault("jupyterhub", {}).setdefault("hub", {}).setdefault(
            "services", {}
        )["hub-health"] = {"apiToken": hub_health_token, "admin": True}
```

I'm guessing, that the hub-health token is the deployer scripts way of granting itself permissions to be able to communicate to the hub health endpoint or something? Or what does it do? I'm not sure. I don't think you would need an api token to ask the hub "are you alive" actually. Is the purpose to ask the hub if its alive?

**Comment by sgibson91** [comment]
> Is the purpose to ask the hub if its alive?

We do ask if it's alive, but we also execute the notebooks in the `deployer/tests` directory, so it needs to start a server too

**Comment by sgibson91** [comment]
Hmmm, but either way we will need to know namespace specifics so it's probably not nicely separable from the deployer. In which case I'll probably leave it as it for now and we can come back to it later if we feel it necessary.

**Comment by consideRatio** [comment]
@sgibson91 makes sense to me!

Regarding the dask-gateway token. We can make the helm chart expose it, but we need to make sure the dask-gateway server could read the jupyterhub api token its given - for example by mounting a k8s secret as a file or environment variable.

I'll want to look into this a bit on how things work currently in the dask-gateway helm chart and dask-gateway-server software which it runs. I'll update this comment to include a link to a issue in dask/dask-gateway i'll create soon.

**Comment by sgibson91** [comment]
Right I have gotten https://github.com/2i2c-org/infrastructure/pull/985 into a state that I believe is mergeable. There are outstanding actions that I would've liked to have tackled in this PR, mainly to make sure docs don't get out of sync, but likely isn't work it given that I'll be away for a week. I'll copy those todos here as well for tracking:

- [ ] Update the deployer to stop relying on fixed filepaths
- [ ] Check the name of the cluster in the name field of the cluster.yaml file matches the folder name. Raise a warning if it doesn't.
- [ ] Update all the "new hub" docs
- [ ] Update the "decomission a hub" issue template

**Comment by sgibson91** [comment]
Also, holy 💩 we run a lot of hubs!

**Comment by sgibson91** [comment]
I opened https://github.com/2i2c-org/infrastructure/issues/1009 to track the breakages to our list of running hubs pages in the docs as the refactor takes place

**Comment by sgibson91** [comment]
### Refactoring proposal

At the minute, the deployer has a lot of hard-coded paths and we are particularly reliant on the `config/clusters` directory existing. To increase flexibility of how we may choose to structure our repo in the future, and make it easier to lift the deployer out into a standalone package, I am going to suggest some changes in convention to our file structure.

1. **The `cluster.yaml` file is the only file the deployer should calculate the filepath for. Other files required to deploy a given hub should be explicitly defined _relative_ to the position of the `cluster.yaml` file.**
   We have already implemented this in a simple fashion. All files under `helm_chart_values_files` are relative to the `cluster.yaml` file they are listed in because they reside in the same folder.
   
2. **We should move config around as needed to fit convention 1 in the most logical manner without traversing a lot of directories up/down where possible.**
   We want to strictly avoid (but not enforce) overly-complicated relative filepaths, such as:
   ```yaml
   helm_chart_values_files:
     - this/is/a/directory/file.values.yaml
     - ../../../../../who/knows/where/this/is/file.values.yaml
   ```
   While the first option could be considered reasonable since the filepath is explicit, the second one can become confusing due to the repetition of the `..` operator. If we find ourselves writing complicated relative filepaths like this, we should reassess and move that config to somewhere more logical.

3. **In the spirit of convention 2, we should no longer house our secrets under a specific folder and mimic directory structure under it. Instead, we should move all config for a cluster, encrypted or not, into the same place and use filenames to track if it should be encrypted and whether is encrypted or not.**
   This convention would involve:
   * Changing `sops` regex from "anything under the `secrets` _directory_ to any _filename_ containing `secret`
   * No longer use the `--in-place` flag for `sops`. Specifically rename an encrypted file to have the `enc-` prefix. This would look something like this:
   ```
   # Encrypting a file
   sops --encrypt file.secret.yaml > enc-file.secret.yaml
   # Decrypting a file
   sops --decrypt enc-file.secret.yaml > file.secret.yaml
   ```
   We can then do two things with the deployer script: 1) identify a file that needs to be handled by `sops` depending on the `enc-` prefix, and 2) check a `enc-`-prefixed file contains the `sops` keys, if not, raise an error that the file may not have been properly encrypted and the secret should be regenerated in case it has been checked into version control. This mimics what `sops` itself does: throws an error if it cannot find the `sops` key when decrypting a file. Some pseudo-code looks like:
   ```
   if FILENAME.startswith("enc-"):
       # Read in file WITHOUT decrypting first
       encrypted_contents = read_yaml(FILENAME)
       if "sops" not in encrypted_contents.keys():
           raise Error(
               "Expecting to find the 'sops' key in this encrypted file but it wasn't found! Please regenerate this secret if it has been checked into version control!"
           )
       else:
           # Decrypt the file and read the decrytped contents
           return decrypted_contents
   else:
       # Do nothing - file should not be encrypted
   ```

Applying these conventions, an example file tree would look like:

```
config
|-- clusters
|   |-- 2i2c
|   |   |-- cluster.yaml  <-- cluster-wide config file
|   |   |-- enc-deployer-credentials.secret.yaml  <-- encrypted secret file
|   |   |-- staging.values.yaml  <-- hub-specific helm chart values for a hub called "staging"
```

**Comment by consideRatio** [comment]
@sgibson91 I love it, all ideas makes sense to me!

I'm adding some motivation for a convention to have `secret` in the filename of files with sensitive content that we want to encrypt, and motivation for also having an `enc-` prefix to indicate a file is encrypted.

1. We can detect these files in general via their filename pattern, and that can be helpful in various situations when we want to give them special treatment without needing to inspect their content etc.
2. We can for example use `.gitignore` to ensure we don't commit non-encrypted secrets. This would be done like this, I think.

   ```gitignore
   # Ignore files (but not directories) with secret in their name
   *secret*
   !*secret*/
   # Make an exception for enc- prefixed files (enc stands for encrypted)
   !enc-*
   # Make an exception for Helm chart templates with secret in their name,
   # they are just templates and are not supposed to contain sensitive content
   # by themselves.
   !**/templates/**/*secret*.yaml
   ```
   
   If you would have decrypted a file to have the same name but without `enc-`, then you could know it would be gitignored, and then clean up with `git clean -Xfd` to delete untracked files (-X), with required confirmation (-f), recursively (-d).

**Comment by GeorgianaElena** [comment]
+1 from me! I like the new suggested struct! Thanks for thinking these things through :blossom: :rocket:

**Comment by choldgraf** [comment]
This looks good to me as well, and it seems reasonable to have all configuration + secrets for a cluster "under one roof".

One question is how this will relate to the multi-cluster spawner work that @yuvipanda has been prototyping. I guess in this case, you would have a `parent cluster` that has the hub itself, and this cluster would contain configuration files for the `spawner clusters` where people could create interactive sessions. Does that sound right?

**Comment by consideRatio** [comment]
> I guess in this case, you would have a parent cluster that has the hub itself, and this cluster would contain configuration files for the spawner clusters where people could create interactive sessions. Does that sound right?

I think the technical changes made as part of this issue is not going to lock into a practice that restricts us later, but rather make them more flexible. If we need new conventions in the future, this is still a nudge in the right direction.

At this point I struggle to carve out conclusions influencing this work by the thought experiment of imagining us maintaining a multicluster spawner deployment.

**Comment by sgibson91** [comment]
IIRC the hub for the multi-cluster spawner doesn't even need to be on a k8s clusters, it can be a plain old VM because the hub part is the only bit it needs to run there.

But the point is, these changes we make now are being made in such a way that we can be flexible about config location in the future without needing to rewrite the deployer _too_ much and also without making decisions about potential future config right now.

**Comment by choldgraf** [comment]
Sounds good to me 👍

**Comment by damianavila** [comment]
+1 on the last @sgibson91 [proposal](https://github.com/2i2c-org/infrastructure/issues/879#issuecomment-1048722191)

**Comment by sgibson91** [comment]
Before we had code that looked like this:

https://github.com/2i2c-org/infrastructure/blob/3d9b57dfdb2ab4d5a53ddb7d555ddf8f7afbb9ed/deployer/hub.py#L564-L572

This ensured that the temporary files still existed when it came to executing the helm upgrade command, but hard-coded how many secret files we were expecting.

I'm not sure how to address this 😕

**Comment by consideRatio** [comment]
@sgibson91 I think you may need an ExistStack. You can't use `with` on a unknown amount of contextmanagers, but you can wrap the unknown amount of contextmanagers.

Here is an example that you could use, I think, I haven't tested this.

```python
@contextmanager
def _get_file_paths_and_decrypt_them_first_if_needed(files):
    """
    This is a context manager that when entered provides a list of
    file paths, where temporary files has been created if needed for
    files that was encrypted and first needed to be decrypted.
    """
    with ExitStack() as stack:
        yield [
            stack.enter_context(verify_and_decrypt_file(f))
            if f.startswith("enc-") or ("secret" in f) else
            f
            for values_file in self.spec["helm_chart_values_files"]
        ]
```

I think this code is a bit ugly, and that perhaps the `verify_and_decrypt_file` function can be remade to allow a non-encrypted file to be passed as well. I figure for example you could have the functions `get_decrypted_file` and `get_decrypted_files`.

```python
@contextmanager
def get_decrypted_files(files):
    """
    This is a context manager that when entered provides a list of
    file paths, where temporary files has been created if needed for
    files that was encrypted and first needed to be decrypted.
    """
    with ExitStack() as stack:
        yield [
            stack.enter_context(get_decrypted_file(f))
            for f in self.spec["helm_chart_values_files"]
        ]
```

**Comment by sgibson91** [comment]
> that perhaps the `verify_and_decrypt_file` function can be remade to allow a non-encrypted file to be passed as well.

It can!

https://github.com/2i2c-org/infrastructure/blob/cc32b3c71b532ee83d3a63939cbeb55afbd6e8af/deployer/utils.py#L125-L130

I wasn't sure this was the right thing to do at the time, hence the comment, but I decided not to break the overall behaviour of the function since the original filepath is also returned if the file does not contain valid JSON/YAML

**Comment by sgibson91** [comment]
I opened https://github.com/2i2c-org/infrastructure/pull/1042 to fix this

**Comment by choldgraf** [comment]
Thanks for all of this awesome discussion and the recent refactoring work!

I took a look at the to do list in the top comment, and think at least one of those things is now completed? I pinged the relevant issue to see if we can close it.

Can we do another audit of the remaining work that we'd like to accomplish here, and update the top comment with a plan? For example:

- Are any other bullet items above close-able?
- Is there anything we want to put off for the future and not finish as a part of this issue / development effort?

**Comment by sgibson91** [comment]
We haven't gotten anywhere near the CI/CD side of things yet, but that is rapidly upcoming.

My opinion is that #279 won't be helpful without #937 having being completed, but they could be happening in parallel to #818

**Comment by sgibson91** [comment]
This is the more granular to-do list I am currently working from https://github.com/2i2c-org/infrastructure/issues/879#issuecomment-1024144966

**Comment by choldgraf** [comment]
That's helpful, thanks! So the question we need to answer then is how to prioritize these three things:

- https://github.com/2i2c-org/infrastructure/issues/937
- https://github.com/2i2c-org/infrastructure/issues/279
- https://github.com/2i2c-org/infrastructure/issues/818

Against other development work that we need to get started soon (for example, setting up BinderHub infrastructure, or improving the story around the configurator+user environment tracking). I would love to hear people's thoughts on how we should prioritize. We can discuss in the meeting on Thursday as well.

**Comment by sgibson91** [comment]
We definitely need #818 - that was the whole point of undergoing this refactoring work!

**Comment by choldgraf** [comment]
So it sounds like you and @consideRatio suggest that we should prioritize https://github.com/2i2c-org/infrastructure/issues/818 over the other projects we could be working on. Is that correct? And if so, can we time-box ourselves so that we have a checkpoint for deciding what to do next? What would be a reasonable estimate to shoot for?

I just want to make sure that this effort is clearly-scoped enough that we are making the right decisions about our own bandwidth and effort, given that there are many other things to do as well.

**Comment by consideRatio** [comment]
If work on these three issues is done in sequence, it should be #818 -> #937 -> #279 I think, but it can also be #818 + #937 in parallel followed by #279 finally.

#818 is absolutely most work.

**Comment by choldgraf** [comment]
If we think it's realistic, then I think it'd be great if we could do

In parallel:

- https://github.com/2i2c-org/infrastructure/issues/818
- https://github.com/2i2c-org/infrastructure/issues/937

and finally

https://github.com/2i2c-org/infrastructure/issues/279


Is that realistic? E.g. could you and @sgibson91 tag-team these? Or would we need another team member to help out with them?


(by the way, apologies if I'm coming across as pedantic here! I am just trying to make sure we're weighing this project against our other options to make sure we're having the most impact!)

**Comment by sgibson91** [comment]
I would love if we could do the tag-team variation on this. But also, I think completing #818 then pausing and coming back to #937 and #279 is plausible. But I don't really wanna give them up completely, because it's really foundational work that will increase the amount of impact we'll have beyond any single BinderHub for a single community.

**Comment by consideRatio** [comment]
I'd like to postpone work #937 until i feel I've managed other work that feels pressing unrelated to 2i2c infra: z2jh/kubespawner/dask-gateway releases and associated changes for those + some jmte maintenance work.

Happy to do #937 at some point in time, but would prefer not going for it at this point in time.

> But I don't really wanna give them up completely, because it's really foundational work that will increase the amount of impact we'll have beyond any single BinderHub for a single community.

I'm not sure if I understand you now. But z2jh, binderhub, and dask-gateway has bundled value file schema's already. For #937 I'm specifically thinking that the charts defined by 2i2c should get schemas for the values as well.

**Comment by sgibson91** [comment]
> I'm not sure if I understand you now.

Sorry. I mean that all the work encapsulated in this project issue is foundational to the impact 2i2c can have because it's related to the reliability, robustness and sustainability of our infrastructure at large. So we can be more successful over a longer time period by investing in this kind of work, compared to the impact deploying a single BinderHub for a single community would have. But I know that's a hard sell to the people handing over their money.

**Comment by choldgraf** [comment]
I think the main benefit of this work is both ourselves (because we will spend less time fighting unexpected fires, or wrapping our heads around confusing config structure) and the communities we serve (because of the reasons @sgibson91 describes). So the prioritization question is basically "how much of an impact will the remaining work have on our own efficiency and robustness".

I think our question should be "where can we spend 10 hours now to save ourselves 100 hours of work later on?"

On the BinderHub side, it's worth noting that we have other BinderHub-style opportunities coming up as well. GESIS would like to fund some BinderHub-style work (https://github.com/2i2c-org/leads/issues/56) and CalTech would also like to collaborate on a BinderHub grant (https://github.com/2i2c-org/leads/issues/62). So this is not a one-off, it is towards building a foundation for running BinderHubs for others as well. (and nonetheless, as @sgibson91 we are contractually obligated to run this BinderHub, so we need to set it up eventually)

**Comment by sgibson91** [comment]
> Does that sound right? It also sounds like folks are leaning towards option 2 above, as long as we track the validation stuff well-enough that we know we'll get to it later.

Yes, I agree this work is important enough to come back to after we address some of the more pressing development issues, and not close it completely.

> I think the reason not to do this is if we think that not validating our charts will potentially generate a big problem for us in the future. But I don't have a great intuition for this right now

Not having this kind of validation makes debugging failed helm upgrades more difficult and involved than it needs to be - that could cover a range of problems from minor to big.

**Comment by consideRatio** [comment]
> Not having this kind of validation makes debugging failed helm upgrades more difficult and involved than it needs to be - that could cover a range of problems from minor to big.

Sometimes just knowing we have validation to catch an issue, makes you not consider that could be an issue - which would save time even though there isn't an issue related to how values were provided.

I love it when my trust in something grows to the extent that I don't need to worry about it. I think perhaps the main benefit of this would be that we can relax a bit with regards to review of changes even though we would also catch mistakes quicker.

**Comment by consideRatio** [comment]
I created #1045 to address both #937 and #279.

**Comment by consideRatio** [comment]
2i2c team meeting notes:
- Sarah plans improvements to start to finish in two weeks

**Comment by sgibson91** [comment]
Hey @2i2c-org/tech-team! The next step of this issue is to provide logic in the deployer that can decide which clusters/hubs need a helm upgrade based off changed filepaths (from a PR, say). I've been experimenting with this idea and have put together a test repository and given you all collaborator access: https://github.com/sgibson91/helm-upgrade-decision-logic

Please feel free to spend some time opening and self-merging PRs to that repo and checking the workflow logs to see if the jobs you were expecting to trigger were indeed triggered! I'll try to write some tests for it as well, and when we're happy it works, I'll migrate the code into the deployer and begin refactoring our `deploy-hubs` workflow.

**Comment by sgibson91** [comment]
I migrated the Python code from the above repo into the following PR https://github.com/2i2c-org/infrastructure/pull/1093

**Comment by damianavila** [comment]
We agreed, at the planning meeting, that this one is now complete!
Congrats @sgibson91 and @consideRatio who were the ones pushing this one forward!!

**Comment by choldgraf** [comment]
wohooo! what an impressive leap forward for our infrastructure! @sgibson91 I wonder if you'd be interested in working with me on a short blog post that explains some of these changes? there were some folks on twitter who were intrigued about our GitHub actions :-)

**Comment by sgibson91** [comment]
@choldgraf yep!

**Comment by sgibson91** [comment]
I opened https://github.com/2i2c-org/team-compass/issues/411 to track the blog post
